### PR TITLE
RBAC: Support github teams #2751

### LINKED ISSUE #2751: RBAC: Support github teams
Currently as of #753 we do support only usernames/organizations roles.

Please upvote if you're interested in roles via gh teams.

```
        - provider: oauth_github
          type: team
          value: "provectus/kafka-backend"
```

### PR DESCRIPTION
**Pull request by maros2710**
- [ ] **Breaking change?** (if so, please describe the impact and migration path for existing application instances)

**What changes did you make?** (Give an overview)
Closes #2751

Added github teams support. (see https://github.com/provectus/kafka-ui/issues/2751)
Now you can set up github team in _rbac_ configuration like this:

```
- provider: oauth_github
  type: team  
  value: "provectus/memelords" #{organization}/{team}
```

**Is there anything you'd like reviewers to focus on?**
I am not very good at reactor, so maybe look at my reactive pipelines, please :)

**How Has This Been Tested?** (put an "x" (case-sensitive!) next to an item)

- [ ] No need to
- [x] Manually (please, describe, if necessary)
- [ ] Unit checks
- [ ] Integration checks
- [ ] Covered by existing automation

**Checklist** (put an "x" (case-sensitive!) next to all the items, otherwise the build will fail)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. **ENVIRONMENT VARIABLES**)
- [x] My changes generate no new warnings (e.g. Sonar is happy)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged

Check out [Contributing](https://github.com/provectus/kafka-ui/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/provectus/kafka-ui/blob/master/CODE-OF-CONDUCT.md)

image

### CODE REVIEW

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/rbac/extractor/GithubAuthorityExtractor.java [hunk 1]
```diff
@@ -98,23 +103,66 @@ public Mono<Set<String>> extract(AccessControlService acs, Object value, Map<Str
...
+                    .flux();
+              }
+              return Flux.empty();
+            })
+        );
+
+    return teams
+        .doOnNext(t -> {
+          if (t.getRight() instanceof Map) {
+            Map<String, String> response = (Map<String, String>) t.getRight();
```
Comment (Haarolean) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/rbac/extractor/GithubAuthorityExtractor.java:144]: We'd like to see a cleaner approach here, let's say this:
1) we filter the roles we have and if there are no subjects matching github teams, we can skip teams fetching altogether
2) the same applies to organizations, if none are present in roles, we don't have to query organizations either

**Review by Haarolean — CHANGES_REQUESTED** [review]
Hi, and thanks for your contribution. Please take a look at the comment I've left.
Also, please refer to our [contributing guide](https://github.com/provectus/kafka-ui/blob/master/CONTRIBUTING.md#working-on-the-issue) and follow one for the future contributions (I've been working on this issue already thus the issue was assigned to me), having you to comment on the issue to grab it beforehand would be nice.

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/rbac/extractor/GithubAuthorityExtractor.java [hunk 2]
```diff
@@ -98,23 +103,66 @@ public Mono<Set<String>> extract(AccessControlService acs, Object value, Map<Str
...
+    Flux<Pair<Role, Object>> teams = Flux.fromIterable(acs.getRoles())
+        .flatMap(role -> Flux.fromIterable(role.getSubjects())
+            .filter(s -> s.getProvider().equals(Provider.OAUTH_GITHUB) && s.getType().equals("team"))
+            .flatMap(subject -> {
+              String orgTeam = subject.getValue();
+              if (orgTeam.contains("/")) {
+                String[] organizationAndTeam = orgTeam.split("/");
+                String org = organizationAndTeam[0];
+                String team = organizationAndTeam[1];
+                WebClient wc = WebClient.create(
```
[the remark this answers is not in the document]
Comment (maros2710) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/rbac/extractor/GithubAuthorityExtractor.java]: I have added git gitHubApiUri property to OAuthProperties.OAuth2Provider that is later used as base GitHub api URI

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/rbac/extractor/GithubAuthorityExtractor.java (hunk 1 above)
Comment (maros2710) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/rbac/extractor/GithubAuthorityExtractor.java:145]: I am not sure, if I understand, can you explain a little bit more please? Thanks!

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/rbac/extractor/GithubAuthorityExtractor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Haarolean) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/rbac/extractor/GithubAuthorityExtractor.java]: We don't need such a property as it should be fetched via OpenID provider configuration request.
You can fetch it via `req.getClientRegistration().getProviderDetails().getUserInfoEndpoint().getUri()`

**Comment by darkLord19** [comment]
@maros2710 are you still interested in completing this or can I try resolving the review comments?

**Comment by Haarolean** [comment]
Closing in favor of #4093

**Pull request closed without merging**